src/client.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  type Middleware,
+  ResponseError,
+  CancellationToken,
+  DidChangeConfigurationNotification,
+  type DidChangeConfigurationRegistrationOptions,
+} from "vscode-languageclient";
+import { Uri } from "vscode";
+import type { PythonExtension } from "@vscode/python-extension";
+
+interface TyMiddleware extends Middleware {
+  isDidChangeConfigurationRegistered(): boolean;
+}
+
+export function createTyMiddleware(pythonExtension: PythonExtension): TyMiddleware {
+  const didChangeRegistrations = new Set<string>();
+
+  const middleware: TyMiddleware = {
+    isDidChangeConfigurationRegistered() {
+      return didChangeRegistrations.size > 0;
+    },
```

Comment:
Is the main reason to keep track of these registrations is that this notification can be dynamically registered by the server? I'm not able to understand what's the benefit of this is.

Reply:
Yes, we should only send the notification if the server asked for it. This ensurs that we don't send this notification to old ty servers not supporting the notification.

Reply:
Hmm, but that requires that ty server would actually need to register for `didChangeConfiguration` request i.e., use the `registerCapability` endpoint to dynamically register for this notification which I wasn't really planning to do (😅) but I think we'll need to now?

Reply:
Is it required to register for `didChangeConfiguration` notification? Doesn't the client always notify the server if there are any changes to the client configuration? We just don't notice it today because VS Code restarts ty. For example, Zed sends the request today but we don't have a handler for it:

```
2025-07-07 15:03:03.894250000  WARN Received notification workspace/didChangeConfiguration which does not have a handler.
```

Reply:
This is not what I would expect from reading the specification:

> This pull model replaces the old push model were the client signaled configuration change via an event. If the server still needs to react to configuration changes (since the server caches the result of workspace/configuration requests) the server should register for an empty configuration change using the following registration pattern:

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration